### ISSUE TITLE
MySQL/MariaDB character set and collation bug

### ISSUE THREAD

**Issue by jeromecc** [prometheusorg/epiClassifier#12]
Processing of non ASCI characters is buggy.

04/09/2017 - 18:37:40,42324>            Processing text: petit corps
�tranger dans l'oeil gauche
04/09/2017 - 18:37:40,42490>                            Global text not
found in 3BT.
04/09/2017 - 18:37:40,42744>                            Term petit found
in Flechies.
04/09/2017 - 18:37:40,42918>                            Term petit corps
not found in Flechies.
04/09/2017 - 18:37:40,43089>                            Term corps found
in Flechies.
04/09/2017 - 18:37:40,43349>NSEpiFlechiesDB::isInDB "corps �tranger" not
found -> Query error 1 Illegal mix of collations
(latin1_swedish_ci,IMPLICIT) and (utf8_general$
04/09/2017 - 18:37:40,43381>                            Term corps
�tranger not found in Flechies.
04/09/2017 - 18:37:40,43548>NSEpiFlechiesDB::isInDB "�tranger" not found
-> Query error 1 Illegal mix of collations (latin1_swedish_ci,IMPLICIT)
and (utf8_general_ci,CO$

Most tables (including flechies) are in Latin1 charset.
Collation was the famous "Latin1 swedish".

Solution: convert all tables to "utf8mb4" with the default collation "utf8mb4_unicode_ci":

ALTER TABLE classer CONVERT TO CHARACTER SET utf8mb4;
ALTER TABLE ... CONVERT TO CHARACTER SET utf8mb4;

Hardcode the default charset for the current connexion with utf8mb4 in nsdatabaseLus/nsSQL.cpp using mysql_set_character_set()

**Comment by jeromecc** [comment]
Test request:
http://localhost:54321/?askfor=getCodesFromText&text=petit+corps+%C3%A9tranger+dans+l%27oeil+gauche
now returns the correct ICPC-2 code:

  <?xml version="1.0" encoding="UTF-8" ?>
  <result>Ok</result><output><codes><code>F76</code></codes></output>
  bye

TODO: add trace log about charset setting success/failure

**Comment by p-ameline** [comment]
The issue is a little bit more complicated than just setting the database character set.
Besides, it is not a bug but a setting.

What occurs is that the C++ program is "transparent" to the character set, hence the real issue is to set the database according to the way the requests are encoded.

For example, the sentence "petit corps étranger dans l'oeil gauche" can get passed as:

UTF8
http://url/?askfor=getCodesFromText&text=petit+corps+%C3%A9tranger+dans+l%27oeil+gauche

or regular
http://url/?askfor=getCodesFromText&text=petit+corps+%E9tranger+dans+l%27oeil+gauche

In the first case (the 'é' char is coded as two digits) the database must converted to UTF8 while in the second case, the best choice is latin1.

**Comment by jeromecc** [comment]
We should probably stick to utf-8 as it is the default character set of GNU/Linux and the web and almost everything.
The future API should deal only with utf-8. I don't know of any serious programming language in use today that cannot deal with utf-8.
I guess it shouldn't be difficult to modify existing implementations to use utf-8 only. All web browsers support utf-8.

Did you try my solution? Is there a problem with it? What is this problem exactly?

What would be the advantage of not getting rid of a legacy character encoding like Latin1?

**Comment by jeromecc** [comment]
About /?askfor=getCodesFromText&text=petit+corps+%C3%A9tranger+dans+l%27oeil+gauche

%C3%A9 is a legacy utf-8 encoding for url but in fact all browsers now accept "é" or any known utf-8 character in the url, particularly in the query.

**Comment by p-ameline** [comment]
The issue here is that the C++ code uses STL std::string. When it sees %E9, it translates this hexa value into a single char. When it sees %C3%A9 it translates these 2 hexa values as two chars.

You can check that E9 is the hexa value for 'é' in the latin1 set (http://cs.stanford.edu/people/mile/iso8859.html) and that U+00E9 is the value for 'é' in UTF16. But, in UTF8, the same char is coded as the two chars c3 a9 (http://www.utf8-chartable.de/).

This is to say that Unicode is not that simple and, given the current 2 bytes representation of the char type, you have to adapt database queries to the way url queries are expressed.

By the way, to have it work on Promethe.org, I kept the UTF8 database but used mysql_set_character_set(_sqlConnector, "latin1"). This to say that we could easily put "latin1" or "utf8" as parameters instead of keeping it hard coded.

**Comment by jeromecc** [comment]
You are confusing bytes and chars. :-)
> The main advantage of UTF-32 is that the Unicode code points are directly indexable. Finding the Nth code point in a sequence of code points is a constant time operation. In contrast, a variable-length code requires sequential access to find the Nth code point in a sequence. This makes UTF-32 a simple replacement in code that uses integers that are incremented by one to identify a character in a string, as was commonly done for ASCII.

One important question: is the current code using this "integers that are incremented by one to identify a character in a string" feature? I guess it does.

If it is using this property, we should use utf-32 (where each glyph is coded by a fixed 4 bytes) to have the same properties than the char object (which is coded with a fixed 1 byte).

The texts we are processing will never be huge, it wouldn't make much sense to ask the ICPC-2 codes of an encyclopedia. So using 4 bytes or 1 byte doesn't change much about RAM or disk usage.

I'm wondering why you cannot use utf-8 only for a project that is compiled and hosted entirely on a Debian VPS... Why can't you get those input strings from your web users in utf-8, convert the MySQL tables to utf8-mb4 and make a connection to the dabase with utf-8?

Maybe the problem comes from your local tests on your Windows machine?

In any case, I don't mind adding 2 parameters for the connection char set and collation in the code using the conf file as long as it defaults to utf-8. ;-)

Free software projects should be inclusive and future-proof. Sticking to Latin1 excludes many people from the project. Even some French words cannot be represented correctly with Latin1. That's why the right French spelling for heart http://localhost:54321/?askfor=getCodesFromText&text=c%C5%93ur (cœur) returns nothing whereas coeur returns a few codes... this is a bug. It affects other medically important French words ("œsophage", ...).

Same for other languages from countries that are working on and using the ICPC (Finish, Dutch).

We should correct this. It will not be so difficult.

**Comment by p-ameline** [comment]
I make no confusion between char and byte. A usual C or C++ char is two bytes long and can be expressed as two hexadecimal digits like E9. A Unicode character can be expressed as multiple chars.

But this is not the problem here. The current C++ code is transparent to this: it just processes a URL as entry string and queries a database in order to know what groups of words are referenced as "access gates to the classification". It doesn't give a shit about character tables internally as far as entries and queries are made with the same character table.

There are multiple standards for URLs. I do not test on a Windows machine but on a Debian server, and the standard coding for URLs has the 'é' being sent as %E9 by the Java code and not a two chars UTF-8 code. Hence the only possible rule here for a component that can be installed in multiple configuration types: have the query character set be settable in the configuration file.

As for the spelling of coeur, it is not "a bug" but a decision based on the fact that the œ character is not present on keyboards, hence people cannot enter it. 
Besides, it would be pretty easy to add the proper spelling to flechies so that cœur and œsophage are treated like coeur and oesophage.

To sum it up. There is no bug... but the need to have those who connect the classification engine between an entry component and a database to be able to set (as a configuration parameter) the proper encoding for database queries so that the global system can work whatever the way the entry component encodes information.

**Comment by jeromecc** [comment]
Done. Still, the Java code running promethe.org should be modified to send out utf8(mb4) encoded data, because it's is the universal standard for Internet & APIs. I just tested https://promethe.org, it doesn't work at all. I will set up a repository for the Java GWT code. Let's fix it together. Website needs https also... @p-ameline
